agent/src/cli/evaluate-autocomplete/Queries.ts

```diff
@@ -0,0 +1,79 @@
+import * as fspromises from 'fs/promises'
+import * as path from 'path'
+
+import Parser, { Query } from 'web-tree-sitter'
+
+import { SupportedLanguage } from '../../../../vscode/src/completions/tree-sitter/grammars'
+
+export type QueryName = 'context'
+
+/**
+ * Queries manages compilation of tree-sitter queries from a directory layout.
+ *
+ * Queries are written in a file structure like this: `LANGUAGE/QUERY_NAME.scm`
+ *
+ * This class caches compilation of queries so that we only read each query once from disk.
+ */
+export class Queries {
+    private cache: CompiledQuery[] = []
+    constructor(private queriesDirectory: string) {}
+    public async loadQuery(parser: Parser, language: SupportedLanguage, name: QueryName): Promise<Query | undefined> {
+        const fromCache = this.cache.find(compiled => compiled.language === language && compiled.queryName === name)
+        if (fromCache) {
+            return fromCache.compiledQuery
+        }
+        return this.compileQuery(parser, language, name)
+    }
+
+    private async compileQuery(
+        parser: Parser,
+        language: SupportedLanguage,
+        name: QueryName
+    ): Promise<Query | undefined> {
+        const languages = [language, ...(grammarInheritance[language] ?? [])]
+        const queryStrings: string[] = []
+        for (const queryLanguage of languages) {
+            const queryPath = path.join(this.queriesDirectory, queryLanguage, `${name}.scm`)
+            try {
+                const stat = await fspromises.stat(queryPath)
+                if (!stat.isFile()) {
+                    continue
+                }
+            } catch {
+                continue
+            }
+            const queryString = await fspromises.readFile(queryPath)
+            queryStrings.push(queryString.toString())
+        }
```

Comment:
Not a big deal but you could fork and join this IO with Promise.all, something like:

```suggestion
        const queryStrings = Promise.all(languages.map(async (language: SupportedLanguage) => {
            const queryPath = path.join(this.queriesDirectory, queryLanguage, `${name}.scm`)
            try {
                const stat = await fspromises.stat(queryPath)
                if (!stat.isFile()) {
                    return
                }
            } catch {
                return
            }
            return (await fspromises.readFile(queryPath)).toString()
        })).filter(id => id);
```

Reply:
This code shouldn't be a performance bottle-neck so I think it's fine to keep it as is. I'm generally not a fan of having large expressions and rely on deep type inference. For example, I think the code can be easier to work with when you already declare that `queryStrings` has type `string[]` because the typechecker can give more detailed error messages when you push another type. It's all subjective in the end